### ISSUE TITLE
Top Posts & Pages widget: applies filter post type choice filter after calculating total posts to display

### ISSUE THREAD

**Issue by macmanx2** [Automattic/jetpack#2369]
The updated Top Posts & Pages widget from Jetpack 3.6 is applying the types choice _after_ the total is collected, which in most cases results in a mismatch with the desired total. This is hard for me to explain with words, so how about an example? :)

http://macmanx.com/ has 1,276 Posts, with stats going back to the day Jetpack was first released, yet when I set my Top Posts & Pages widget to display my top 10 posts, it only displays 6. If I set it to display my top 10 posts _and_ pages, it displays 10, and 4 of those are pages.

In the above example, rather than fetching my top 10 posts, the widget fetches my top 10 posts _and_ pages, then filters out the 6 posts from that.

**Comment by jeherve** [comment]
Also reported here:
https://wordpress.org/support/topic/top-posts-and-pages-widget-displays-too-many-entries-when-filtered-by-page-types?replies=1&view=all

**Comment by kraftbj** [comment]
One suggestion for this is to pull a higher number (whatever reasonable... 20, 50, etc), then filter them locally to return the specific number of posts or pages.

**Comment by kraftbj** [comment]
Also noting that this would also be resolved by #2196, though my earlier suggestion would work depending on the timeline for the rewrite.

**Comment by scriptgeni** [comment]
Top Posts & Pages (JetPack) stopped working, I had 10 posts displaying yesterday, now it's only showing one. What can be wrong?

**Comment by jeherve** [comment]
@scriptgeni Can you please use Jetpack's Debugger module to send us some more information about your site?

1) Go to the Jetpack page in your dashboard.
2) Click on the Debug link appearing at the bottom of the page.
3) Click the link that says "click here to contact Jetpack support."
4) Fill in the description box and your name and email address.
5) Click the "Contact Support" button.

Thanks!

**Comment by atrus1701** [comment]
I recently fixed the issue #2196 mentioned above.  I can tell you that in resolving that issue, it did not fix this issue.  This is due to the way Top Posts are retrieved from the server.

Issue 1:
The Top Posts endpoint on the REST API does not support filtering by post type.  So if the widget is setup to view the top posts of a not-often-viewed custom post type, those posts might be low on the overall list.  Meaning, it possibly could not show up on the retrieved list of 10 or so posts.  You could try reconnecting and retrieving more posts, but it would be better to fix it on the server end, in my opinion.

Issue 2:
In addition, the REST endpoint restricts the max number of posts per day and doesn't retrieve an overall list of top posts.  This probability won't affect most sites with the default option of top posts over 2 days, but if a site wants the top posts over a longer period, such as a month, this could allow a post that is visited more often over time to be cut off by posts that were only popular for a short time.

After looking through the public-api code, to fix the first issue would require a change to the select statement used to retrieve the list to something like `SELECT from $pageviews_table JOIN $site_posts_table WHERE post_type IN ($post_types)`.  However, for the second issue, the select statement is more complex with either a temp table or `UNION ALL` needed to collect a full list of posts.  Of course, you could always run multiple select statements, merge the results, and filter using PHP as well.  But I don't have the full view of how the database structure of WP.com is setup so there might be an easier way,

**Comment by jeherve** [comment]
Also reported here:
https://wordpress.org/support/topic/top-posts-category-filter-top-posts-limit-question?replies=1&view=all

**Comment by csonnek** [comment]
Also reported here: 
https://wordpress.org/support/topic/popular-posts-widgets?replies=1

**Comment by annezazuu** [comment]
reported in 2699464-t

**Comment by t22james** [comment]
I'll just add some weight to this being a pretty killer 'bug'. It leaves the user experience of this plugin really really frustrating, where the options which you expose in the widget, in a sense and in practice, seem utterly broken. When I ask for 10 posts, of post type X, If that is a low-view post type, I will get 0 items displaying, even though I know there are posts of that type and I know they have views. Of course this is still 'logical' because you used the word 'max' posts, but it also means that the plugin is all but useless for anything other than a standard top 5 of anyposts/page. Which is fine if that's all it was sold as, but which is again, so frustrating when the UI suggests that the plugin could be so much more powerful. 

In parting can i suggest that the ideal functionality for usability of the plugin would be:

Specify number of posts (exact!): X (filled in as much as possible via your stats system, and then padded with extra/fallback posts if not enough via stats)
Specify post type: Y (only caveat here is if I ask for 10 posts, but my chosen post type literally only has 6 posts published, you obviously couldn't show 10)

I do appreciate the work gone into this plugin, and it seems so close to being really really awesome, if only we could get more logical, predictable results!

Thanks

**Comment by jeherve** [comment]
Also reported here:
https://wordpress.org/support/topic/exclude-an-entire-tags-from-top-posts?replies=5&view=all#post-8655589

**Comment by t22james** [comment]
Seeing you have this set for 4.4 milestone, but can't seem to find what version you're currently at? we have a client waiting on this and after a month of waiting we're at the point ofneeding to switch to another system if there's no movement :( any ideas when it might be out?

**Comment by jeherve** [comment]
@t22james We're about to release 4.2. This issue is indeed scheduled for 4.4, for some of the reasons outlined above.

Until the problem is fixed, you could try using the `jetpack_top_posts_widget_count` filter to change the number of posts being displayed:
https://developer.jetpack.com/hooks/jetpack_top_posts_widget_count/

However, you would still run into the same issue; the number you would pick with the filter wouldn't be the final number; post types you won't want to display in the widget would then be withdrawn from that number.

Another alternative would be to use the filter above to query for a large number of posts (20 for example), and then use the `jetpack_widget_get_top_posts` filter to limit the number of displayed posts to 10:

```php
function jeherve_more_top_posts() {
	return '20';
}
add_filter( 'jetpack_top_posts_widget_count', 'jeherve_more_top_posts' );

function jeherve_only_ten_top_posts( $posts, $post_ids, $count ) {
	return array_splice( $posts, 10 );
}
add_filter( 'jetpack_widget_get_top_posts', 'jeherve_only_ten_top_posts', 10, 3 );
```

That should solve your issue.

**Comment by t22james** [comment]
@jeherve hey thanks for the update and and attempt at temporary fix, but I think that still suffers the same problem, in that both of those filters run BEFORE the set of posts is filtered by post type. So again we can have a set of 10 returning from a larger set of 20 (as your snippet implies), but then this code runs:

(around line 279 of top-posts.php)
```
// Filter the returned posts. Remove all posts that do not match the chosen Post Types.
if ( isset( $types ) ) {
	foreach ( $posts as $k => $post ) {
		if ( ! in_array( $post['post_type'], $types ) ) {
			unset( $posts[$k] );
		}
	}
}
```

which then knocks those 10 down to potentially 0. I think that is back to the OP issue.

**Comment by jeherve** [comment]
That's a good point, my bad, I should have double checked at what point `jetpack_widget_get_top_posts` was applied.

You could still use that filter, though, and do the post type filtering there. Once you've done that, you'll have a list of posts matching your selection, and you can then check the number of posts in the array, and make sure you only have 10 posts. You could then have it return nothing if it doesn't have enough posts to fit your needs.

**Comment by t22james** [comment]
@jeherve yep had that idea, then went to sleep ;) but here is a snippet I've ended up with for now:

For anyone coming across this issue, try this out... this will return an exact number of posts by using the posts returned by top posts plugin, padded out by some 'latest' posts if not enough top posts were found. NOTE that this requires hard coding post-types and post count, which will override settings you have in the admin widget itself. If you have different widgets with different settings, you'll need to add some if statements and replicate them in the code below.

FYI this should become redundant when this ticket is resolved.
```
<?php 

// force up the number of posts requested
function jeherve_more_top_posts() {
    return '100';
}
add_filter( 'jetpack_top_posts_widget_count', 'jeherve_more_top_posts' );


// alter the set of posts returned in top posts
function jeherve_relimit_topposts_displayed_count( $posts, $post_ids, $count ) {
	
	// define the post types you want to include in your widget
	$type_of_posts = array('post', 'custom_post1', 'custom_post2');

	// define how any posts to display (WILL display this number, so long as you have this many posts in your DB).
	$num_of_posts = 5;
	
	// create a set of posts, using top posts plugin, and padding with latest posts from your chosen post types, if not enough top posts were found
	$posts = create_better_topposts($posts, $type_of_posts, $num_of_posts, $offset);

    return $posts;
}
add_filter( 'jetpack_widget_get_top_posts', 'jeherve_relimit_topposts_displayed_count', 10, 3 );

// our custom function to supplement the native top posts query
function create_better_topposts($posts, $type_of_posts, $num_of_posts){

	// remove any posts which don't match our desired set
	// we do this here, so that the existing filter in top posts plugin, which runs AFTER $count is applied, is pre-empted, and thus we will get our exact number of desired posts.
	foreach ( $posts as $k => $post ) {
        if ( !in_array($post['post_type'], $type_of_posts) ) {
            unset( $posts[$k] );
        }
    }

    // reindex posts and count
    $posts = array_values($posts);
	$remaining_post_count = count($posts);

	// remove surplus any posts
    if($remaining_post_count > $num_of_posts){

		$posts = array_splice( $posts, $num_of_posts );

	// else, pad with extra posts (latest of desired post type/s) if not enough popular posts were returned
    } else if($remaining_post_count < $num_of_posts){
    	
    	//make array of post ids we already have from 'most popular', so we don't double up
    	$existing_post_ids = array_map(function($post) {
    	    return $post['post_id'];
    	}, $posts);

    	$padding_posts_required = ($num_of_posts - $remaining_post_count);

    	$args = array(
    		'posts_per_page'   => $padding_posts_required,
    		'orderby'          => 'date',
    		'order'            => 'DESC',
    		'post__not_in'     => $existing_post_ids,
    		'post_type'        => $type_of_posts,
    		'post_status'      => 'publish',
    		'ignore_sticky_post‌'=> true
    	);

		// query extra posts
    	$padding_posts = get_posts( $args );

		// conform the new posts into our existing top-posts-plugin friendly array
    	foreach ( $padding_posts as $padding_post ) {

    		// Both get HTML stripped etc on display
    		if ( empty( $padding_post->post_title ) ) {
    			$title_source = $padding_post->post_content;
    			$title = wp_html_excerpt( $title_source, 50 );
    			$title .= '&hellip;';
    		} else {
    			$title = $padding_post->post_title;
    		}

    		$post_id = $padding_post->ID;

    		$permalink = get_permalink( $post_id );

    		$post_type = $padding_post->post_type;

    		$posts[] = compact( 'title', 'permalink', 'post_id', 'post_type' );
    	}

    }

    return $posts;
}
```

**Comment by t22james** [comment]
@jeherve hmm though am i right that this line (471):
```
$post_view_posts = stats_get_csv( 'postviews', array( 'days' => absint( $days ), 'limit' => 11 ) );
```
will still be limiting the pool of posts to 11? though curiously when I am printing the `$posts` that is being passed into `jetpack_widget_get_top_posts` filter, i only seem to ever have an array with 10 items. so actually even the fix above, though it helps, still can never increase the pool of popular posts available above that hardcoded 11....

**Comment by t22james** [comment]
ok, so final code where I'm now pretty much re-querying everything, to bypass all hardcoded result limits. This will ensure you get as close as popular to a true list of popular posts, while still having a fallback to latest posts. See other notes in my previous code block re: usage.

add to functions.php, and define $type_of_posts, $num_of_posts and $days
```
function jeherve_more_top_posts() {
    // force up the number of posts requested (this bit may be redundant now)
    return '100';
}
add_filter( 'jetpack_top_posts_widget_count', 'jeherve_more_top_posts' );


// alter the set of posts returned in top posts
function jeherve_relimit_topposts_displayed_count( $posts, $post_ids, $count ) {

    // define the post types you want to include in your widget - this will override your settings in admin widget
    $type_of_posts = array('post', 'custom_post1', 'custom_post2');

    // define how any posts to display (WILL display this number, so long as you have this many posts in your DB) - this will override your setting in admin widget
    $num_of_posts = 5;

    // days of data to use
    $days = 3;

    // create a set of posts, using top posts plugin, and padding with latest posts from your chosen post types, if not enough top posts were found
    $posts = create_better_topposts($posts, $type_of_posts, $num_of_posts, $days);

    return $posts;
}
add_filter( 'jetpack_widget_get_top_posts', 'jeherve_relimit_topposts_displayed_count', 10, 3 );

// our custom function to supplement the native top posts query
function create_better_topposts($posts, $type_of_posts, $num_of_posts, $days){

    //replicating code from top-posts plugin code so that we can increase the number of results returned from their hardcoded 11, to 150... ensures we get more actual popular posts
    $post_view_posts = stats_get_csv( 'postviews', array( 'days' => absint( $days ), 'limit' => 150 ) );
    if ( ! $post_view_posts ) {
        return array();
    }
    $post_view_ids = array_filter( wp_list_pluck( $post_view_posts, 'post_id' ) );
    if ( ! $post_view_ids ) {
        return array();
    }

    // overwrite jetpack posts, as we are reconstructing from scratch. crucial difference is we arn't limiting results to $count or $num posts here, we we'll only do that below, after filtering out post types
    $posts = array();

    foreach ( (array) $post_view_ids as $post_id ) {
        $post = get_post( $post_id );

        if ( ! $post )
            continue;

        // hide private and password protected posts
        if ( 'publish' != $post->post_status || ! empty( $post->post_password ) || empty( $post->ID ) )
            continue;

        // Both get HTML stripped etc on display
        if ( empty( $post->post_title ) ) {
            $title_source = $post->post_content;
            $title = wp_html_excerpt( $title_source, 50 );
            $title .= '&hellip;';
        } else {
            $title = $post->post_title;
        }

        $permalink = get_permalink( $post->ID );

        $post_type = $post->post_type;

        $posts[] = compact( 'title', 'permalink', 'post_id', 'post_type' );
    }

    // remove any posts which don't match our desired set
    // we do this here, so that the existing filter in top posts plugin, which runs AFTER $count is applied, is pre-empted, and thus we will get our exact number of desired posts.
    foreach ( $posts as $k => $post ) {
        if ( !in_array($post['post_type'], $type_of_posts) ) {
            unset( $posts[$k] );
        }
    }

    // reindex posts and count
    $posts = array_values($posts);
    $remaining_post_count = count($posts);

    // remove surplus any posts
    if($remaining_post_count > $num_of_posts){

        array_splice( $posts, $num_of_posts );

    // else, pad with extra posts (latest of desired post type/s) if not enough popular posts were returned
    } else if($remaining_post_count < $num_of_posts){

        //make array of post ids we already have from 'most popular', so we don't double up
        $existing_post_ids = array_map(function($post) {
            return $post['post_id'];
        }, $posts);

        $padding_posts_required = ($num_of_posts - $remaining_post_count);

        $args = array(
            'posts_per_page'   => $padding_posts_required,
            'orderby'          => 'date',
            'order'            => 'DESC',
            'post__not_in'     => $existing_post_ids,
            'post_type'        => $type_of_posts,
            'post_status'      => 'publish',
            'ignore_sticky_post‌'=> true
        );

        // query extra posts
        $padding_posts = get_posts( $args );

        // conform the new posts into our existing top-posts-plugin friendly array
        foreach ( $padding_posts as $padding_post ) {

            // Both get HTML stripped etc on display
            if ( empty( $padding_post->post_title ) ) {
                $title_source = $padding_post->post_content;
                $title = wp_html_excerpt( $title_source, 50 );
                $title .= '&hellip;';
            } else {
                $title = $padding_post->post_title;
            }

            $post_id = $padding_post->ID;

            $permalink = get_permalink( $post_id );

            $post_type = $padding_post->post_type;

            $posts[] = compact( 'title', 'permalink', 'post_id', 'post_type' );
        }

    }

    return $posts;
}
```

**Comment by chjan** [comment]
Hopefully I am not diluting this thread, but I have a request to the widget "Top post and pages" and this is the option to exclude (not show in the list) certain posts by their id. Possible?

**Comment by jeherve** [comment]
@chjan You can use the `jetpack_widget_get_top_posts` filter to do that. You can check the documentation here:
https://developer.jetpack.com/hooks/jetpack_widget_get_top_posts/

**Comment by chjan** [comment]
Thanks but this will give me a maintenance problem unless you can explain how to do this once only?

 

Please add to the gui of this widget. It will take you 5min of coding. Thanks!

 

Kind regards / Vennlig hilsen

 

Carl Henrik Janson

+47 951 048 00

**Comment by jeherve** [comment]
@chjan We most likely won't add a new field to exclude post IDs in the widget settings, as some site owners could find it confusing;  Post IDs is not something that's very clear for everyone.

Could you tell me more about what you'd like to achieve? Do you find yourself having to exclude different posts or pages often? If so, what are those pages? Do they have anything in common? Why would you like to exclude them?

**Comment by chjan** [comment]
I have a few comments:

 

Any feature that is consider complex is easily hidden out of clear sight under the “advanced features” button, possible including a state ment that “this is nothing for you unless you are competent on post id etc”.

 

I have a few particular postings that are gaining in popularity because they pinpoint bugs in audio drivers of Realtek and products of LG. This is because there are links to my explanations in various support groups with a lot of people striving with the same as I did. I am a technician and has been able to describe remedies, but: These types of posts are bordering on being uninteresting as a greatest hits filing, at least in the long run (and they keep on gaining in popularity since the large companies do not fix their bugs).

 

Thus, I want this out of the list, but in a way that does not create a need for manual maintenance along the lines of your suggestion yesterday. Hope you understand.

 

While we are on the line: There are features that have moved in fairly advanced versions under normal WP to too simple versions in Jetpack, one example is the option to blog directly from an (in my case) Outlook email. To be able to write in a professional editor the first draft of a post, it is a good thing. I use this editor now and many times each day. It is significantly better than the tinyMCE with all its annoying immaturities, so to speak. Now I cannot use this feature anymore because the solution in Jetpack lacks complexity and feature sets as compared to the old version.

 

Generally speaking, please do not diminish functionality just like that. Below please find the requirement spec for posting from outlook, this template is not at all grasped by Jetpack. I am thus now working  from word, which is more or less equal to Outlook editor. Unfortunately, MS lives still in the old world and has done they same as you do, diminished the feature set as measured against their “in the attic” Windows Live Writer. The missing link is that it is impossible to write a page page in Word. So if you are creating a page-page, then you need to write it as a blog, then install a plugin that can convert to and from post-pages and page-pages. 

 

It is cumbersome, annoying, timeconsuming, etc, and the whole reason is that the feature set has been diminished down to something only partly usable. One could ask why such a pioneer in editing has fallen so deep down in the mud, but it has to do with incapacity to investigate and analyse user needs.

 

Not bragging but I was the inventor of FreePad, indeed a forerunner to today’s iPad type of devices, so the whole of me is focusing on how to exploit technology so that people can be more effective. The layer between the human hand and the processor…. It is very important, and it is quite an art to satisfy both simplicity and complexity needs by the users. Indeed!

 

This email is a timethief on my day plan, but if it can make you think about my need to filter away certain postings, then it will be worth it.

 

Hope you can take this as an inspiration :) 

 

I have also one idea: Create a jetpack “thingy” that enables cut and paste of images directly in a post while editing online, and storing it properly according to media library rules. (And make sure it works also for the template in red below.). Bluntly speaking, people are extremely adaptive to diminished feature sets. This cut and paste has been available in offline editors for decades, but then we all go online only to experience the “crap” of not being able to cut and paste images. Amazing!

 

Kind regards / Vennlig hilsen

 

Carl Henrik Janson

+47 951 048 00

 

 

THE OLD TEMPLATE THAT I HAVE AS A SIGNATURE IN OUTLOOK AND THAT MAKE ME EXTREMELY EFFICIENT WHEN I NEED TO SUDDENLY JOT DOWN A POST. HOPE TO GET IT BACK (also for page-pages).

 

 

[title Epost blogg test]

[category 1p1]

[excerpt]some excerpt[/excerpt]

[comments off]

[status draft]

 

Blogg tekst

 

 

[end] Info I resten av epost blir ikke synlig

[slug some-url-name]

[password secret-password]

[slideshow]

[nogallery]

[more]

[nextpage]

[geotag on | off]

[publicize off|yahoo|twitter|facebook]

[category Hol,Fo]

[tags one potato, two potato, three potato, more]

[publicize off] 

[publicize twitter]

[publicize twitter]my new post[/publicize]

[delay +1 hour]

[delay +2 days]

**Comment by jeherve** [comment]
> Any feature that is consider complex is easily hidden out of clear sight under the “advanced features” button

That's not an approach we usually take, mainly because we try to follow 3 key elements of WordPress' philosophy:
- Design for the Majority
- Decisions, not Options
- Clean, Lean, and Mean
- Striving for Simplicity
You can read more about those principles [here](https://wordpress.org/about/philosophy/).

If an advanced setting is useful to very few users, we try to provide filters, and documentation to use those filters.

In your example, you could use that filter to automatically remove all posts belonging to a specific category on your site. If you think posts about bugs in audio drivers of Realtek and products of LG shouldn't be in the Top Posts Widget, you can use the filter to remove all those posts at once.

> one example is the option to blog directly from an (in my case) Outlook email.

That feature is [Post By Email](https://jetpack.com/support/post-by-email/), and it's still part of Jetpack. 2 of the shortcodes you mentioned may not always work well, as per #146 and #4834, but the rest should work just fine. 
If you experience issues with that feature, I'd recommend that you contact our support team [via this contact form](http://jetpack.com/contact-support/) so we can investigate and understand why it doesn't work on your site.

> Create a jetpack “thingy” that enables cut and paste of images directly in a post while editing online, and storing it properly according to media library rules. (And make sure it works also for the template in red below.).

You can use this plugin to enable this feature in WordPress:
https://wordpress.org/plugins/image-elevator/

I haven't tested it myself, but it seems to do just what you're looking for.

Modern email clients like Gmail, and WordPress's post editor also support drag and drop to add images to your posts in a few seconds.

**Comment by chjan** [comment]
3 more comments then.

 

1.     I think the post per email now lacks the option to publish as draft and that is a showstopper, but I am uncertain if I remember correctly. There was something that made me stop using it and move over to MS Word. I now have MS Word in my fingers for “email type of blogging”. You lost a customer.

 

2.     Do you provide a solution that maintains my personal customization in the form of e.g. your suggested function to filter particular posts out of the list? If not, I am stuck with yet another plugin that can handle that (if possible), or to reinsert code on each new version of jetpack, not?

 

3.     I do not agree that it is good practice to remove features that are mature for many years, just because users are too few. It would be better to teach WP students how to blog from email (for example), it is an excellent effective way of writing posts.

 

Kind regards 

 

Carl Henrik Janson

+47 951 048 00

**Comment by jeherve** [comment]
> I think the post per email now lacks the option to publish as draft and that is a showstopper, but I am uncertain if I remember correctly.

I just ran a quick test, it works 👍 

> I think the post per email now lacks the option to publish as draft and that is a showstopper, but I am uncertain if I remember correctly.

Don't hesitate to give it another try, and [send us an email](http://jetpack.com/contact-support/) if you find something that doesn't work. We'll help you figure things out!

> Do you provide a solution that maintains my personal customization in the form of e.g. your suggested function to filter particular posts out of the list? If not, I am stuck with yet another plugin that can handle that (if possible), or to reinsert code on each new version of jetpack, not?

You don't need to edit the Jetpack plugin files. In fact, I would strongly recommend against editing plugin files in general, if only for the reason you mentioned: your customizations disappear every time you update the plugin.

Filters like the one I mentioned above allow you to customize the behaviour of a plugin, a theme, or WordPress itself, all from outside of that plugin / theme. You can read more about how to use filters here:
https://developer.jetpack.com/customize-jetpack-actions-filters/

In practice, the easiest way to add that code snippet to your site would be to use a functionality plugin like [this one](https://wordpress.org/plugins/code-snippets/), that will allow you to manage snippets from your dashboard.

**Comment by jeherve** [comment]
Also reported here:
https://wordpress.org/support/topic/popular-posts-shortcode-display-count/

**Comment by jeherve** [comment]
Also reported here:
https://wordpress.org/support/topic/bug-jetpack-top-posts-pages-of-posts-displayed-incorrect/

**Comment by lschuyler** [comment]
Reported in 3165769-t

**Comment by AndreaBarghigiani** [comment]
I really appreciate all the work that this community is doing in order to improve the plugin, but I have a doubt: is this issue fixed?

I am asking because as reading from [here](https://github.com/Automattic/jetpack/issues/2369#issuecomment-239804925) it should be fixed by the 4.4 version and now we're running the 5.1 but I still have the same problem.

I still have to try the fix proposed by @t22james from [this topic on](https://github.com/Automattic/jetpack/issues/2369#issuecomment-240278046), but before I do it I would like to understand a bit more what is going on.

Can we help the team testing some code? Shall we use the solution already proposed?

I am asking because I think this is a bug that needs to be fixed because otherwise, we as developers will have a problem working with this widget, especially if we have clients that need some sort of "top post by category".

I already tried to load more posts and then filter them out, but it does not assure to have a fixed number of posts inside our list. 

Sometime the list could be even empty and it's difficult to understand how to run a check to hide the title widget if the list is empty.

Thank you for your hard work and for the support as well.

Talk you soon,
Andrea

**Comment by jeherve** [comment]
> is this issue fixed?

No, this issue is still open, we don't have any perfect solution yet, so nothing for you to test either. We'll update this issue once we have some news!

**Comment by Ipstenu** [comment]
Yep, still an issue.

If Jetpack an detect the number of public post types, then why not just collect a number of posts (per the widget) and multiply it by the available post types.

So if I have posts, pages, and portfolios, and my widget is set to show the top 5, then it would collect 15 posts.

**Comment by jeherve** [comment]
Related: 7973-wpcom

**Comment by tmmbecker** [comment]
also reported here 1006113-zen

**Comment by macmanx2** [comment]
This appears to be a variation of this bug: https://wordpress.org/support/topic/bug-popular-posts-and-pages-widget/

**Comment by khoug** [comment]
> This appears to be a variation of this bug: https://wordpress.org/support/topic/bug-popular-posts-and-pages-widget/

Thanks for your reply, let me check the code, i go try fix it and i say here.

**Comment by darnelldibbles** [comment]
Another report here:
https://wordpress.org/support/topic/top-posts-widget-not-working-correctly/

**Comment by ntpixels** [comment]
Also reported by 2471335-zen

**Comment by jeherve** [comment]
This should be fixed as of #15109.
